unstructured/chunking/title.py

```diff
@@ -198,59 +198,89 @@ def _split_elements_by_title_and_table(
     combine_text_under_n_chars: int,
     new_after_n_chars: int,
     max_characters: int,
-) -> List[List[Element]]:
-    sections: List[List[Element]] = []
-    section: List[Element] = []
-
-    for i, element in enumerate(elements):
-        metadata_differs = False
-        if i > 0:
-            last_element = elements[i - 1]
-            metadata_differs = _metadata_differs(
-                element.metadata,
-                last_element.metadata,
-                include_pages=not multipage_sections,
-            )
+) -> Iterator[_TextSection | _TableSection | _NonTextSection]:
+    """Implements "sectioner" responsibilities.
+
+    A _section_ can be thought of as a "pre-chunk", generally determining the size and contents of a
+    chunk formed by the subsequent "chunker" process. The only exception occurs when a single
+    element is too big to fit in the chunk window and the chunker splits it into two or more chunks
+    divided mid-text. The sectioner never divides an element mid-text.
+
+    The sectioner's responsibilities are:
+
+        * **Segregate semantic units.** Identify semantic unit boundaries and segregate elements on
+          either side of those boundaries into different sections. In this case, the primary
+          indicator of a semantic boundary is a `Title` element. A page-break (change in
+          page-number) is also a semantic boundary when `multipage_sections` is `False`.
 
-        section_length = sum([len(str(element)) for element in section])
+        * **Minimize chunk count for each semantic unit.** Group the elements within a semantic unit
+          into sections as big as possible without exceeding the chunk window size.
 
-        new_section = (
-            (section_length + len(str(element)) > max_characters)
-            or (isinstance(element, Title) and section_length > combine_text_under_n_chars)
-            or (metadata_differs or section_length > new_after_n_chars)
+        * **Minimize chunks that must be split mid-text.** Precompute the text length of each
+          section and only produce a section that exceeds the chunk window size when there is a
+          single element with text longer than that window.
+    """
```

Comment:
What is the behavior when the last element is less than `combine_text_under_n_chars` (or if the only element is less than that)? We might want to add a test for that.

Reply:
Not sure what you're asking exactly, do you mean "last" like "preceding" or like "terminal element".

But elements don't really come into combination, only whole sections do. This is what the next stacked PR fixes, so I'd look at those tests and see if you think I've missed something.

So this PR doesn't try to change combination behavior or improve the tests for it, that is left to the next PR.